### PR TITLE
Improve cim ETS member inference completion

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Improves the member type inference for cim based commands by adding a fallback to the short namespaceless name if a full name was specified.

## PR Context
Many of the CDXML modules uses the `.types.ps1xml` type files to add additional members. In most cases they use a short name that doesn't specify the WMI namespace like: `Microsoft.Management.Infrastructure.CimInstance#MSFT_NetFirewallRule` but they still declare the full type with the namespace in the outputtype attribute: `Microsoft.Management.Infrastructure.CimInstance#root/standardcimv2/MSFT_NetFirewallRule`  

This is completely valid as the ciminstance type adapter adds both the full and short name to the list of typenames: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/cimSupport/other/ciminstancetypeadapter.cs#L234  
So the type inference code also needs to handle this.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [X] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): 
- **User-facing changes**
  - [X] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: 
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: 
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: 
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 1]
```diff
@@ -195,7 +195,14 @@ internal IList<object> GetMembersByInferredType(PSTypeName typename, bool isStat
                 // Look in the type table first.
                 if (!isStatic)
                 {
-                    var consolidatedString = new ConsolidatedString(new[] { typename.Name });
+                    // The Ciminstance type adapter adds the full typename with and without a namespace to the list of type names.
+                    // So if we see one with a full typename we need to also get the types for the short version.
+                    // For example: "CimInstance#root/standardcimv2/MSFT_NetFirewallRule" and "CimInstance#MSFT_NetFirewallRule"
+                    Match match = Regex.Match(typename.Name, "(Microsoft\\.Management\\.Infrastructure\\.CimInstance#)(.+\\/)(.+$)");
```
Comment (jborean93) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:201]: I'm unsure if it's worth it but potentially you could avoid regex by seeing if it starts with `Microsoft.Management.Infrastructure.CimInstance#` then splitting the value after with `/` to build the short type name entry.

There's also a source generator for regex to have it compile the pattern automatically for you that may be of interest https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators?pivots=dotnet-8-0.

Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:201]: It is better not use regex (if possible) for security reasons.

Comment (MartinGC94) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:201]: Sure, I'll remove the regex usage. What's the security problem with regex though?

Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:201]: > What's the security problem with regex though?

I see the Regex is in "parser". So we must check if input data are checked enough to exclude _infinite_ Regex, for example.

**Review by SeeminglyScience — APPROVED** [review]
LGTM, thanks Martin!

**Pull request merged**